{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a2c0bb48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # A00_environment_and_genome_setup =============================================\n",
    "\n",
    "# # in terminal, set up snm3C pipeline environment\n",
    "# module load anaconda3 # or otherwise activate conda\n",
    "# conda env create -f Documentation/snm3Cseq_taurus.yml # follow setup instructions\n",
    "\n",
    "# # then, run the following commands\n",
    "# qsub Scripts/A00a_genome_dl_index.sub  # ‡\n",
    "# qsub Scripts/A00b_genome_prep_bismark.sub\n",
    "# qsub Scripts/A00c_annotations_bed.sub # ‡\n",
    "\n",
    "# # ‡ fast enough to run interactively"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6978ac0d",
   "metadata": {},
   "source": [
    "## minimal project directory setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c092f422",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # make sure that the project directory & minimal file structure exists\n",
    "# # by running in terminal / bash\n",
    "# dir_proj=/u/project/cluo_scratch/chliu/IGVF_iPSC_snm3Cseq_YZCL47\n",
    "# mkdir $dir_proj; cd $dir_proj\n",
    "# mkdir Metadata Notebooks Scripts\n",
    "\n",
    "# # note that if running pipeline via .ipynb notebooks,\n",
    "# # hitting \"run all\" on notebooks will fill up the Scripts folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8225a9d",
   "metadata": {},
   "source": [
    "## snm3C_parameters (crucial to review!)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "23fe91be",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../snm3C_parameters.env\n",
    "\n",
    "# parameters file --------------------------------------------------------------\n",
    "# note: filepaths are relative to project directory or absolute paths\n",
    "# recommend using absolute filepaths for all \"ref_dir\"/genome related files by\n",
    "# find & replace \"/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda\" --> \n",
    "#        folder where you want your genome ref_dir to reside\n",
    "\n",
    "# primary analysis/project folder\n",
    "dir_proj=/u/project/cluo_scratch/chliu/IGVF_iPSC_snm3Cseq_YZCL47\n",
    "\n",
    "# folder with raw data\n",
    "# (.fastq, for our group usually split across 4 lanes of a novaseq run)\n",
    "dir_originalfastq=/u/project/cluo/Shared_Datasets/source_fastq/yzcl47\n",
    "\n",
    "# scratch folder if available, otherwise can set be same as dir_proj\n",
    "dir_scratch=/u/project/cluo_scratch/chliu\n",
    "\n",
    "# reference genome/files\n",
    "# these are generated after running scripts A00a-A00c\n",
    "# alt: if a Hoffman2 user with access to our partition, can use other organisms in my Genomes dir\n",
    "# (versions also exist +/- lambda spike-in : below is an IGVF hg38 + lambda)\n",
    "ref_dir=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda\n",
    "ref_dir_bowtie1=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/bowtie1\n",
    "ref_fasta=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/IGVF_GRCh38_plus_lambda.fa\n",
    "ref_gtf=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/gencode.v43.chr_patch_hapl_scaff.annotation.gtf\n",
    "\n",
    "\n",
    "\n",
    "# genome reference parameters  -------------------------------------------------\n",
    "# point to $ref_dir directory, no need to change final file name\n",
    "# (i.e., keep \"chromsizes.tsv\") besides organism\n",
    "# again may be generated after running A00a-A00c\n",
    "\n",
    "# reference genome\n",
    "ref_chromsizes=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/chromsizes.tsv\n",
    "ref_flat=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/annotations/refFlat.txt.gz \n",
    "ref_rrna=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/annotations/rRNA.intervallist \n",
    "\n",
    "# gene features\n",
    "ref_genebody=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/annotations/genebody.bed\n",
    "ref_geneslop2k=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/annotations/geneslop2k.bed\n",
    "ref_exon=/u/project/cluo/chliu/Genomes/IGVF_hg38_pluslambda/annotations/exon.bed\n",
    "\n",
    "\n",
    "\n",
    "# other miscellaneous parameters -----------------------------------------------\n",
    "# OK as-is / defaults don't have to be changed\n",
    "\n",
    "# metadata files\n",
    "metadat_plate=Metadata/A01b_plate_metadata.csv\n",
    "metadat_well=Metadata/A01c_well_filepath.csv\n",
    "\n",
    "# fastqc (A02c, A03c)\n",
    "wells_to_run=Scripts/A02c_random_fastqc_wells.txt\n",
    "numwells_run=2\n",
    "overwrite_random_wells=\"false\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52e88a95",
   "metadata": {},
   "source": [
    "## (A00a) download reference .fasta, index, get chrom sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0778a67b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# note: if using UCLA Hoffman2 cluster, can use my reference genomes\n",
    "# just check that the \"bowtie1\" folder is present\n",
    "# (see filepaths listed above)\n",
    "\n",
    "# there may be extraneous files for genotyping pipelines that don't interfere with mCT or m3C\n",
    "# (.fa.gz --> .bwt, .pac, .ann, .amb, .sa, .fai, .dict; generated with below commands)\n",
    "# bwa index GRCh38.primary_assembly.genome.fa.gz\n",
    "# gatk CreateSequenceDictionary -R GRCh38.primary_assembly.genome.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "af0d41f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00a_genome_dl_index.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00a_genome_dl_index.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00a_genome_dl_index\n",
    "#$ -l h_rt=8:00:00,h_data=8G\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snm3Cseq_taurus # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# download/extract ref ---------------------------------------------------------\n",
    "\n",
    "cd ${ref_dir}\n",
    "\n",
    "# double check filepaths and hard-coded names here # <--\n",
    "# download IGVF consortium-specified .fa & .gtf as example (hard coded below)\n",
    "# see https://lh3.github.io/2017/11/13/which-human-reference-genome-to-use\n",
    "wget https://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/001/405/GCA_000001405.15_GRCh38/seqs_for_alignment_pipelines.ucsc_ids/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna.gz # <--\n",
    "wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_43/gencode.v43.chr_patch_hapl_scaff.annotation.gtf.gz # <--\n",
    "\n",
    "# extract compressed files as needed\n",
    "# put final, uncompressed names here into \"snm3C_parameters.env\" (as \"ref_fasta\" and \"ref_gtf\")\n",
    "for gzfile in *gz\n",
    "do\n",
    "    gunzip -c ${gzfile} > ${gzfile/.gz/}\n",
    "done\n",
    "\n",
    "# note: if ref_fasta doesn't have .fa or .fasta file format\n",
    "# bismark may have issues detecting it in ${ref_dir}\n",
    "# rename .fna --> .fa, for example\n",
    "for f in *fna; do mv ${f} ${f/fna/fa}; done\n",
    "\n",
    "\n",
    "\n",
    "# account for Lambda spike-in --------------------------------------------------\n",
    "\n",
    "# note on working with Lambda phage or other spike-in (bisulfite conversion efficiency QC)\n",
    "# add to reference genome at this step. comment out below four lines--\n",
    "# Escherichia phage Lambda, complete genome (GenBank: J02459.1) via NCBI\n",
    "\n",
    "# wget \"https://www.ncbi.nlm.nih.gov/sviewer/viewer.cgi?tool=portal&save=file$=seqview&db=nuccore&report=fasta&id=215104\" -O lambda.fa\n",
    "# sed -i \"1s/.*/>chrL/\" lambda.fa\n",
    "# cat GCA_000001405.15_GRCh38_no_alt_analysis_set.fa lambda.fa > IGVF_GRCh38_plus_lambda.fa \n",
    "# rm GCA_000001405.15_GRCh38_no_alt_analysis_set.fa\n",
    "\n",
    "\n",
    "\n",
    "# index, chrom sizes -----------------------------------------------------------\n",
    "\n",
    "# extract bp length/chromosome\n",
    "samtools faidx ${ref_fasta}\n",
    "cut -f 1-2 ${ref_fasta}.fai > chromsizes.tsv\n",
    "\n",
    "# .fa --> .dict\n",
    "picard CreateSequenceDictionary -R ${ref_fasta}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00a_bwa_index' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "016fe42d",
   "metadata": {},
   "source": [
    "## (A00b) prep bismark genome ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4b0ee576",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00b_genome_prep_bismark.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00b_genome_prep_bismark.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00b_prep_bismark\n",
    "#$ -l h_rt=24:00:00,h_data=8G\n",
    "#$ -pe shared 4\n",
    "#$ -hold_jid A00a_genome_dl_index\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -----------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snm3Cseq_taurus # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# bismark index -----------------------------------------------------------------\n",
    "\n",
    "# if using TAURUS version of m3C pipeline (\"snm3Cseq_taurus\")\n",
    "# use bowtie1 to index the genome\n",
    "\n",
    "mkdir ${ref_dir}/bowtie1\n",
    "cd ${ref_dir}/bowtie1\n",
    "\n",
    "ln -s ${ref_fasta} input.fa\n",
    "\n",
    "if [[ ! -s Bisulfite_Genome ]]\n",
    "then\n",
    "    bismark_genome_preparation $PWD --bowtie1\n",
    "fi\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00b_genome_prep_bismark' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6ea0359",
   "metadata": {},
   "source": [
    "## (A00c) extract annotations from gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "699c6112",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00c_gtf_annotations_bed.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A00c_gtf_annotations_bed.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -N A00c_gtf_annotations_bed\n",
    "#$ -l h_rt=0:30:00,h_data=8G\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -----------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snm3Cseq_taurus # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# process .gtf --> .bed ------------------------------------------------------\n",
    "\n",
    "mkdir ${ref_dir}/annotations\n",
    "\n",
    "python Scripts/A00c_gtf_annotations_bed.py\n",
    "\n",
    "\n",
    "\n",
    "# UCSC/Picard utilities for QC metrics ---------------------------------------\n",
    "\n",
    "cd ${ref_dir}\n",
    "\n",
    "# rRNA interval list\n",
    "picard BedToIntervalList -I annotations/rRNA.bed -O \\\n",
    "    annotations/rRNA.intervallist -SD ${ref_fasta/fa/dict}\n",
    "\n",
    "# ref flat\n",
    "gtfToGenePred -genePredExt -geneNameAsName2 ${ref_gtf} refFlat.tmp.txt\n",
    "cut -f 1-10 refFlat.tmp.txt > refFlat.tmp1\n",
    "cut -f 12 refFlat.tmp.txt > refFlat.tmp2\n",
    "\n",
    "paste refFlat.tmp1 refFlat.tmp2 > annotations/refFlat.txt\n",
    "gzip annotations/refFlat.txt\n",
    "\n",
    "rm refFlat.tmp*\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A00c_gtf_annotations_bed' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e65bab1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A00c_gtf_annotations_bed.py\n",
    "\n",
    "# ==============================================================================\n",
    "# A00c_gtf_annotations_bed.py \n",
    "# exports four annotation-related files to $ref_dir (reference genome)\n",
    "# for mcds creation & down-stream analysis\n",
    "# ==============================================================================\n",
    "\n",
    "# recommend running interactively in python/Jupyter to check outputs\n",
    "# works for .gtfs from GENCODE, but double check fmt for other sources\n",
    "\n",
    "# # if running interactively, need to load some lines from snm3C_parameters.env\n",
    "# # os.environ['ref_dir'] = \"/u/project/cluo/chliu/Genomes/IGVF_hg38\" or the below loop\n",
    "# # (use absolute versus relative path of parameters.env file if below not working!)\n",
    "# envvar_needed = ['dir_proj', 'ref_dir', 'ref_gtf', 'ref_chromsizes']\n",
    "# try:\n",
    "#     os.environ['ref_dir']\n",
    "# except KeyError:\n",
    "#     envspec = pd.read_csv(\"../snm3C_parameters.env\", sep = \"=\", comment=\"#\", header = None\n",
    "#                ).set_axis(['varname', 'varpath'], axis = 1\n",
    "#                ).query('varname in @envvar_needed')\n",
    "#     for index, row in envspec.iterrows():\n",
    "#         os.environ[row[\"varname\"]] = row[\"varpath\"]\n",
    "# os.chdir(os.environ['dir_proj'])\n",
    "\n",
    "\n",
    "# load packages ----------------------------------------------------------------\n",
    "\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "\n",
    "\n",
    "# load reference info ----------------------------------------------------------\n",
    "\n",
    "os.chdir(os.environ['ref_dir'])\n",
    "os.makedirs(\"annotations/\", exist_ok=True)\n",
    "\n",
    "gtf_file = pd.read_csv(os.environ['ref_gtf'],\n",
    "                       comment = \"#\", delimiter=\"\\t\", header = None)\n",
    "chrom_sizes = pd.read_csv(os.environ['ref_chromsizes'], sep = \"\\t\", header = None)\n",
    "chrom_sizes.columns = ['#chr', 'chrlen'] \n",
    "\n",
    "\n",
    "\n",
    "# genebody ---------------------------------------------------------------------\n",
    "# .gtf to .bed (1-based --> 0 based)\n",
    "# columns: chr, start, end, ENSG identifier\n",
    "# note that this contains mitochondrial, ribosomal, lncRNAs, etc;\n",
    "# this may or may not be desireable in downstream analyses\n",
    "\n",
    "genebody = gtf_file[gtf_file.iloc[:, 2] == 'gene'].iloc[:, [0, 3, 4, 8]]\n",
    "genebody.iloc[:, 1] = genebody.iloc[:, 1] - 1 # start changes to 0-pos\n",
    "genebody.columns = ['#chr', 'start', 'end', 'annot']\n",
    "genebody.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# extract info from the annot column (;, \")\n",
    "genebody['gene'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[1])\n",
    "genebody['symbol'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[5])\n",
    "genebody['type'] = genebody['annot'].transform(lambda x: str(x).split('\\\"')[3])\n",
    "\n",
    "# .gtf checks: should be zero\n",
    "if sum(genebody.gene.duplicated()) != 0 | sum(genebody.start >= genebody.end) != 0:\n",
    "  print(\"WARNING: check .bed outputs; was gene info was processed correctly from .gtf?\")\n",
    "\n",
    "# export ENSG --> Symbol\n",
    "genebody.drop('annot', axis = 1, inplace = True)\n",
    "genebody.to_csv(\"annotations/ensembl_to_symbol.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "# bed4 format for .allcools\n",
    "# (drops strand info, which can be useful for other software)\n",
    "genebody = genebody.iloc[:, 0:4]\n",
    "genebody.to_csv(\"annotations/genebody.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "\n",
    "# gene +/- 2kb -----------------------------------------------------------------\n",
    "# above, but padding a 2kb region\n",
    "# past manuscripts do this for higher mC modality coverage, but less interpretable\n",
    "\n",
    "g2k = genebody.copy()\n",
    "g2k.iloc[:, 1] = g2k.iloc[:, 1] - 2000\n",
    "g2k.iloc[:, 2] = g2k.iloc[:, 2] + 2000\n",
    "\n",
    "# but check +/-2kb still within chromosome length\n",
    "# (low # genes affected, but may cause downstream issues)\n",
    "g2k.loc[g2k.start < 0, 'start'] = 0\n",
    "\n",
    "g2k = pd.merge(g2k, chrom_sizes, on = '#chr') \n",
    "filter_chrlen = g2k.end > g2k.chrlen\n",
    "g2k.loc[filter_chrlen, 'end'] = g2k.chrlen[filter_chrlen]\n",
    "g2k.drop('chrlen', axis = 1, inplace=True)\n",
    "g2k.to_csv(\"annotations/geneslop2k.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "\n",
    "# exonic -----------------------------------------------------------------------\n",
    "# exon-level annotations, useful for STAR exon-only quant, potential ASE\n",
    "\n",
    "exon = gtf_file[gtf_file.iloc[:, 2] == 'exon'].iloc[:, [0, 3, 4, 8]]\n",
    "exon.iloc[:, 1] = exon.iloc[:, 1] - 1 # start changes to 0-pos\n",
    "exon.columns = ['#chr', 'start', 'end', 'annot']\n",
    "\n",
    "exon['gene'] = exon['annot'].transform(lambda x: str(x).split('\\\"')[1])\n",
    "exon['transcript'] =  exon['annot'].transform(lambda x: str(x).split('\\\"')[3])\n",
    "exon = exon.drop('annot', axis = 1).reset_index(drop=True)\n",
    "exon.to_csv(\"annotations/exon.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "\n",
    "# rRNA genes -------------------------------------------------------------------\n",
    "# for some QC metrics after RNA alignments\n",
    "\n",
    "rRNA = genebody.loc[gtf_file.iloc[:, 8].str.contains(\"rRNA\"), :]\n",
    "rRNA.to_csv(\"annotations/rRNA.bed\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
