{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # A05_quantify_mC overall cmds ===============================================\n",
    "\n",
    "# qsub Scripts/A05a_bam2allc.sub  # † \n",
    "# qsub Scripts/A05b_check_allcs.sub\n",
    "# qsub Scripts/A05c_global_mC_stats.sub # †\n",
    "# qsub Scripts/A05d_allc_to_mcds.sub # *\n",
    "\n",
    "# * = job array based on \"platenum\"\n",
    "# † = job array based on \"batchnum\" (two rows at a time)\n",
    "# ‡ fast enough to run interactively"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05a_bam2allc.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05a_bam2allc.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=4:00:00,h_data=24G\n",
    "#$ -N A05a_bam2allc\n",
    "#$ -t 1-512\n",
    "#$ -hold_jid_ad A04c_coverage\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snm3Cseq_taurus # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "skip_complete=true # <-- for help with incomplete jobs\n",
    "overwrite_partial=true # <-- for help with incomplete jobs\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "# (generally not customizeable because output names set by bismark)\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A04a_dir_bismark\"))\n",
    "\n",
    "bam_in=($(query_metadat \"A04a_bam_final\"))\n",
    "\n",
    "# allc.tsv.gz by default\n",
    "allc_out=($(query_metadat \"A05a_allc\"))\n",
    "allctbi_out=($(query_metadat \"A05a_allctbi\"))\n",
    "\n",
    "\n",
    "\n",
    "# print target files -----------------------------------------------------------\n",
    "\n",
    "echo \"batch number: $SGE_TASK_ID\"\n",
    "echo \"processing the following rows in well metadata file ($metadat_well):\"\n",
    "for row in ${target_well_rows[@]}\n",
    "    do\n",
    "        echo -e \"$row\\t${wellprefix[$row]}\"\n",
    "    done\n",
    "echo -e \"\\n\\n\"\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run mapping .bam\n",
    "    cd $dir_proj\n",
    "    \n",
    "    if [[ -s ${allc_out[$row]} \\\n",
    "        && -s ${allctbi_out[$row]} ]]\n",
    "    then\n",
    "        if [ $(wc -c < ${allc_out[$row]}) -le 50 ];\n",
    "        then echo \".allc exists, but seems to be empty. re-running '${wellprefix[$row]}'.\"\n",
    "        else\n",
    "            echo -e \"final .allc for '${wellprefix[$row]}' already exists.\"\n",
    "            if [[ \"${skip_complete}\" == \"true\" ]]\n",
    "            then\n",
    "                echo \"skip_complete == true. skipping this well.'\"\n",
    "                continue\n",
    "            else\n",
    "                echo \"skip_complete != true. re-running this well.'\"\n",
    "            fi\n",
    "        fi\n",
    "    fi\n",
    "    \n",
    "    if [[ ! -e ${bam_in[$row]} ]]\n",
    "    then\n",
    "        echo -e \"\\n\\n.bam for '${wellprefix[$row]}' missing. skipping this well.\"\n",
    "        continue\n",
    "    fi\n",
    "\n",
    "    # .bam --> .allc (generally 3-5min)\n",
    "    echo -e \"\\n\\ncoverting .bam to .allc for '${wellprefix[$row]}'...\\n\\n\"\n",
    "    cd ${dir_proj}/${dir_well[$row]}\n",
    "\n",
    "    if [ $(ls allc.tsv.gz.temp.* | wc -l) > 0 ]\n",
    "    then\n",
    "        echo \"some intermediate .allc files found in directory; removing.\"\n",
    "        rm allc.tsv.gz.temp.*\n",
    "    fi\n",
    "    \n",
    "    allcools bam-to-allc -bam ${dir_proj}/${bam_in[$row]} \\\n",
    "         --reference_fasta $ref_fasta --output_path ${dir_proj}/${allc_out[$row]} \\\n",
    "         --convert_bam_strandness --cpu 1\n",
    "\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05a_bam2allc' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05b_check_allcs.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05b_check_allcs.$JOB_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=2:00:00,h_data=4G\n",
    "#$ -N A05b_allccheck\n",
    "#$ -hold_jid A05a_bam2allc\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs)\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              } \n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "check_filepaths_in_assay() {\n",
    "    for file in $@\n",
    "        do \n",
    "        if [[ ! -s $file ]]\n",
    "            then\n",
    "                echo \"missing '$file'\"\n",
    "            fi\n",
    "        done\n",
    "}\n",
    "\n",
    "check_filepath_by_batch() {\n",
    "target_array=($@)\n",
    "batches_to_rerun=()\n",
    "for ((target_batch=1; target_batch<=nbatches; target_batch++))\n",
    "    do\n",
    "        target_well_rows=()\n",
    "        for ((row=1; row<=nwells; row++))\n",
    "        do\n",
    "            if [[ \"${batchnum[$row]}\" == \"${target_batch}\" ]]\n",
    "            then\n",
    "                target_well_rows+=($row)\n",
    "            fi\n",
    "        done\n",
    "\n",
    "        batch_file_list=${target_array[@]: ${target_well_rows[0]}:${#target_well_rows[@]} }\n",
    "    \n",
    "        num_files_missing=$(check_filepaths_in_assay ${batch_file_list[@]} | wc -l)\n",
    "\n",
    "        if [[ ${num_files_missing} > 0 ]]\n",
    "        then\n",
    "            batches_to_rerun+=(${target_batch})\n",
    "            echo -e \"${target_batch} \\t ${num_files_missing}\"\n",
    "        fi\n",
    "    done \n",
    "    \n",
    "    if [[ ${#batches_to_rerun[@]} > 0 ]]\n",
    "    then\n",
    "        echo \"batches to re-run:\"\n",
    "        echo \"${batches_to_rerun[*]}\"        \n",
    "    fi\n",
    "}\n",
    "\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "\n",
    "nwells=${#batchnum[@]}\n",
    "nbatches=${batchnum[-1]}\n",
    "\n",
    "\n",
    "\n",
    "# apply checks for A04a output -------------------------------------------------\n",
    "\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "echo \"A. printing number of final .bams missing (by batch)... \"\n",
    "echo \"-----------------------------------------------------------------\"\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "allcfile=($(query_metadat \"A05a_allc\"))\n",
    "allctbi=($(query_metadat \"A05a_allctbi\"))\n",
    "\n",
    "echo \"checking .allc.tsv.gz\"\n",
    "echo -e \"batchnum\\tnum_missing\"\n",
    "check_filepath_by_batch ${allcfile[@]}\n",
    "\n",
    "echo \"checking .allc.tsv.gz.tbi\"\n",
    "check_filepath_by_batch ${allctbi[@]}\n",
    "\n",
    "\n",
    "echo -e \"\\n\\nsuggest re-running and checking sublog output of above batches.\"\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"B. checking each expected .bam file (from $metadat_well)\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "echo -e \"\\nchecking .allc.tsv.gz:\\n\"\n",
    "check_filepaths_in_assay ${allcfile[@]}\n",
    "\n",
    "echo -e \"\\nchecking .allc.tsv.gz.tbi:\\n\"\n",
    "check_filepaths_in_assay ${allctbi[@]}\n",
    "\n",
    "\n",
    "echo -e \"\\n* checks the A05a output columns of 'metadat_well' if the file exists and is non-empty.\"\n",
    "echo \"* if none missing, will only output target column names above.\"\n",
    "echo \"* if some declared 'missing' but all other checks OK, may just be no/few reads surviving trimming.\"\n",
    "echo \"  (check 'fastq_demultip/' and associated fastp logs e.g., fastq_trimmed/wellprefix.html report)\"\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n-----------------------------------------------------------------\"\n",
    "echo \"C. checking log files for issues.\"\n",
    "echo -e \"-----------------------------------------------------------------\\n\"\n",
    "\n",
    "echo \"checking if 'completed' in sublogs/A05a_bam2allc* output.\"\n",
    "echo \"if any filename is printed, the associated batch may have not completed allc gen.\"\n",
    "\n",
    "grep -c 'ended on' sublogs/A05a_bam2allc* | awk -F \":\" '$2==0 {print $1}'\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05b_allccheck' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05c_global_mC_stats.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05c_global_mC_stats.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=6:00:00,h_data=4G\n",
    "#$ -N A05c_global_mC_stats\n",
    "#$ -t 1-512\n",
    "#$ -hold_jid_ad A05a_bam2allc\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:   \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init -------------------------------------------------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "action_metadata_exists=\"overwrite\" # <-- for help with incomplete jobs; overwrite, append, rename\n",
    "check_lambda=true # <-- lambda phage spike-in (check if included in wet lab prep)\n",
    "check_autosomal_only=false # <-- exclude mitochondrial, scaffolds, lambda\n",
    "\n",
    "# output file\n",
    "# (should probably make these explicitly named vs batchnumbered in future)\n",
    "metadat_out=Metadata/A04d_mCfrac_${SGE_TASK_ID}.tsv\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths -----------------------------------------------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' $metadat_well\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "batchnum=($(query_metadat \"batchnum\"))\n",
    "nwells=${#batchnum[@]}\n",
    "\n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${batchnum[$row]}\" == \"$SGE_TASK_ID\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "allcfile=($(query_metadat \"A05a_allc\"))\n",
    "allctbi=($(query_metadat \"A05a_allctbi\"))\n",
    "\n",
    "# for methylation fraction\n",
    "calc_mC_frac () {\n",
    "    awk -v context=$1 '$4 ~ context' ${2:-${tmpfile}} |\n",
    "        awk '{mC+=$5; tot+=$6} END {\n",
    "            if (tot == 0) { print \"0\\t0\\tNA\" } else { print mC \"\\t\" tot \"\\t\" mC/tot }}'\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# print target files -----------------------------------------------------------\n",
    "\n",
    "echo \"batch number: $SGE_TASK_ID\"\n",
    "echo \"processing the following rows in well metadata file ($metadat_well):\"\n",
    "for row in ${target_well_rows[@]}\n",
    "    do\n",
    "        echo -e \"$row\\t${wellprefix[$row]}\"\n",
    "    done\n",
    "echo -e \"\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "# checking for existing metadata file ------------------------------------------\n",
    "\n",
    "# columns are wellprefix, then for a given sequence context\n",
    "# num reads supporting methylation, total coverage, then fraction methylated\n",
    "\n",
    "header=\"wellprefix\\tmLam\\tLam\\tmLamfrac\\tmCCC\\tCCC\\tmCCCfrac\\tmCG\\tCG\\tmCGfrac\\tmCH\\tCH\\tmCHfrac\\n\"\n",
    "\n",
    "if [[ ! -e ${metadat_out} ]]\n",
    "then\n",
    "    printf ${header} > ${metadat_out}\n",
    "else\n",
    "    echo \"WARNING: ${metadat_out} seems to already exist.\"\n",
    "    if [[ $action_metadata_exists == \"overwrite\" ]]\n",
    "        then\n",
    "        echo \"overwriting the existing file. (action_metadata_exists=='overwrite').\"\n",
    "        printf \"${header}\" > ${metadat_out}\n",
    "    elif [[ $action_metadata_exists == \"append\" ]]\n",
    "        then\n",
    "            echo \"appending to the existing file. (action_metadata_exists=='append')\"\n",
    "    elif [[ $action_metadata_exists == \"rename\" ]]\n",
    "        then\n",
    "            metadat_out=Metadata/A04d_mCfrac_${SGE_TASK_ID}.tsv\n",
    "            echo \"renaming output file to ${metadat_out} to avoid overwriting (action_metadata_exists=='rename').\"\n",
    "            printf \"${header}\" > ${metadat_out}\n",
    "    else\n",
    "        echo \"exiting. (check 'action_metadata_exists' variable if to change action.)\"\n",
    "        exit 1\n",
    "    fi\n",
    "fi\n",
    "\n",
    "\n",
    "\n",
    "# loop through allc, calculate methylation fracs -------------------------------\n",
    "# assuming methylation given for CHN or CGN contexts\n",
    "# (usually <1 minute per file)\n",
    "\n",
    "for row in ${target_well_rows[@]} \n",
    "do\n",
    "\n",
    "    # check for existing mapping output\n",
    "    # if final outputs exist, skip; else run awk tabulization\n",
    "    cd ${dir_proj}\n",
    "    \n",
    "    if [[ ! -s ${allcfile[$row]} \\\n",
    "        && ! -s ${allctbi[$row]} ]]\n",
    "    then\n",
    "        echo -e \"WARNING: final .allc files for '${wellprefix[$row]}' missing. skipping this well.'\"\n",
    "    else\n",
    "    \n",
    "    echo \"processing '${wellprefix[$row]}'...\"\n",
    "        \n",
    "    # create temporary unzipped .allc file to awk through\n",
    "    # (if there are \"tmp_*\" files in '${dir_proj}/Metadata', this script may have failed/timeout)\n",
    "    tmpfile=\"Metadata/tmp_${wellprefix[$row]}\"\n",
    "    if [[ \"$check_lambda\" == \"true\" ]]\n",
    "    then\n",
    "        # use a partially methylated lambda, methylated at all but CAG, CTG\n",
    "        gunzip -c ${allcfile[$row]} | awk '$1 == \"chrL\" && $4 != \"CAG\" && $4 != \"CTG\"' \\\n",
    "                > ${tmpfile}_lambda\n",
    "        mlambda=$(calc_mC_frac \"^[ACTG][ACTG][ACTG]\" ${tmpfile}_lambda )\n",
    "        rm ${tmpfile}_lambda\n",
    "    else\n",
    "        mlambda=\"0\\t0\\tNA\"\n",
    "    fi\n",
    "    \n",
    "    if [[ \"$check_autosomal_only\" == \"true\" ]]\n",
    "        then\n",
    "            # warning: may not capture scaffolds named e.g., chr6_SCAFF\n",
    "            gunzip -c ${allcfile[$row]} | awk '$1 ~ \"^chr[0-9]\"' > ${tmpfile}\n",
    "        else\n",
    "            gunzip -c ${allcfile[$row]} | awk '$1 != \"chrL\"' > ${tmpfile}\n",
    "    fi\n",
    "\n",
    "    # calculate CCC, CGN, CHN\n",
    "    mccc=$( calc_mC_frac \"^CCC\" )\n",
    "    mcg=$( calc_mC_frac \"^CG[ACTG]\" )\n",
    "    mch=$( calc_mC_frac \"^C[ACT][ACTG]\" )\n",
    "\n",
    "    # record metrics, remove tmp file\n",
    "    echo -e \"${wellprefix[$row]}\\t${mlambda}\\t${mccc}\\t${mcg}\\t${mch}\" \\\n",
    "        >> ${metadat_out}\n",
    "    rm ${tmpfile}\n",
    "\n",
    "fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05c_global_mC_stats' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:   \" `date `\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "cat > ../Scripts/A05d_allc_to_mcds.sub\n",
    "\n",
    "#!/bin/bash\n",
    "#$ -cwd\n",
    "#$ -o sublogs/A05d_allc2mcds.$JOB_ID.$TASK_ID\n",
    "#$ -j y\n",
    "#$ -l h_rt=24:00:00,h_data=8G,highp\n",
    "#$ -N A05d_allc2mcds\n",
    "#$ -t 1-32\n",
    "#$ -pe shared 8\n",
    "#$ -hold_jid A05a_bam2allc\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:    \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID started on:    \" `date `\n",
    "echo \" \"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# environment init ------------------—------------------—-----------------------\n",
    "\n",
    ". /u/local/Modules/default/init/modules.sh # <--\n",
    "module load anaconda3 # <--\n",
    "conda activate snm3Cseq_taurus # <--\n",
    "\n",
    "export $(cat snm3C_parameters.env | grep -v '^#' | xargs) # <--\n",
    "\n",
    "\n",
    "\n",
    "# extract target filepaths ------------------—------------------—---------------\n",
    "\n",
    "# helper functions\n",
    "query_metadat () {\n",
    "  awk -F',' -v targetcol=\"$1\" \\\n",
    "      'NR==1 {\n",
    "                for (i=1;i<=NF;i++) {\n",
    "                    if ($i==targetcol) {assayout=i; break} }\n",
    "                print $assayout\n",
    "              }\n",
    "      NR>1 {\n",
    "                print $assayout\n",
    "            }' ${metadat_well}\n",
    "}\n",
    "\n",
    "# extract target wells, print values for log\n",
    "platenum=($(query_metadat \"platenum\"))\n",
    "nwells=${#platenum[@]}\n",
    "  \n",
    "target_well_rows=()\n",
    "for ((row=1; row<=nwells; row++))\n",
    "do\n",
    "    if [[ \"${platenum[$row]}\" == \"${SGE_TASK_ID}\" ]]\n",
    "    then\n",
    "        target_well_rows+=($row)\n",
    "    fi\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# filepaths associated with target rows in well-level metadata -----------------\n",
    "# (generally not customizeable because output names set by bismark)\n",
    "\n",
    "wellprefix=($(query_metadat \"wellprefix\"))\n",
    "dir_well=($(query_metadat \"A04a_dir_bismark\"))\n",
    "\n",
    "bam_in=($(query_metadat \"A04a_bam_mergedsort\"))\n",
    "\n",
    "allc_out=($(query_metadat \"A05a_allc\"))\n",
    "allctbi_out=($(query_metadat \"A05a_allctbi\"))\n",
    "\n",
    "\n",
    "\n",
    "# print target files -----------------------------------------------------------\n",
    "\n",
    "echo \"batch number: $SGE_TASK_ID\"\n",
    "echo \"processing the following rows in well metadata file ($metadat_well):\"\n",
    "\n",
    "for row in ${target_well_rows[@]}\n",
    "do\n",
    "    echo -e \"${row}\\t${wellprefix[$row]}\"\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "# make .tsv of allcs -----------------------------------------------------------\n",
    "# will appear as Metadata/A05d_allclist_*\n",
    "\n",
    "tsv_target_allcs=Metadata/A05d_allclist_${SGE_TASK_ID}.tsv\n",
    "if [[ -s ${tsv_target_allcs} ]]\n",
    "then\n",
    "    rm ${tsv_target_allcs}\n",
    "fi\n",
    "\n",
    "for row in ${target_well_rows[@]}\n",
    "do\n",
    "    if [[ -s ${allc_out[$row]} && -s ${allctbi_out[$row]} ]]\n",
    "    then\n",
    "        echo -e \"${wellprefix[$row]}\\t${allc_out[$row]}\"  >> ${tsv_target_allcs}\n",
    "    fi\n",
    "done\n",
    "echo -e \"\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "# run mcds generation ----------------------------------------------------------\n",
    "\n",
    "if [[ ! -s mcds ]]\n",
    "then\n",
    "    mkdir mcds\n",
    "fi\n",
    "\n",
    "# .allc files --> aggregated into mcds regions\n",
    "# note: the mcds is not a single file, but more like a directory with binary compression\n",
    "# if this step fails may leave a tmp file in the mcds/ folder\n",
    "allcools generate-dataset  \\\n",
    "    --allc_table ${tsv_target_allcs} \\\n",
    "    --output_path mcds/${SGE_TASK_ID}.mcds \\\n",
    "    --chrom_size_path ${ref_chromsizes} \\\n",
    "    --obs_dim cell \\\n",
    "    --cpu 8 \\\n",
    "    --chunk_size 400 \\\n",
    "    --regions chrom100k 100000 \\\n",
    "    --regions genebody ${ref_genebody} \\\n",
    "    --quantifiers chrom100k count CGN,CHN \\\n",
    "    --quantifiers genebody count CGN,CHN\n",
    "\n",
    "# not standard, but but consider adding:\n",
    "#    --regions chrom5k 5000 \\\n",
    "#    --regions geneslop2k $ref_geneslop2k \\\n",
    "# (genebody +/- 2kb to include promoter region + increase cov)  \n",
    "\n",
    "\n",
    "\n",
    "echo -e \"\\n\\n'A05d_allc2mcds' completed.\\n\\n\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:    \" `hostname -s`\n",
    "echo \"Job $JOB_ID.$SGE_TASK_ID ended on:    \" `date `\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
